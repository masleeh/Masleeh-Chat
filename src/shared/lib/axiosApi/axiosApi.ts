import axios from 'axios'

const $api = axios.create({
    withCredentials: true,
    baseURL: import.meta.env.VITE_API_URL
})

$api.interceptors.request.use((config) => {
    config.headers.Authorization = `Bearer ${localStorage.getItem(import.meta.env.VITE_LOCALSTORAGE_TOKEN_KEY)}`
    return config
})

$api.interceptors.response.use((config) => {
    return config   
}, async (error) => {
    const originalRequest = error.config;
    if (error.response.status === 401 && error.config && !error.config._isRetry) {
        originalRequest._isRetry = true
        try {
            const response = await axios.get(`${import.meta.env.VITE_API_URL}/auth/refresh`, { withCredentials: true })
            localStorage.setItem(import.meta.env.VITE_LOCALSTORAGE_TOKEN_KEY, response.data.access_token)
            return $api.request(originalRequest)
        } catch (error) {
            console.log("Not authorized")
        }
    }
    throw error
})

export default $api