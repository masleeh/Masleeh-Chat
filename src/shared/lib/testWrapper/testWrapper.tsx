import { render } from "@testing-library/react"
import { IStateSchema, TestStoreProvider } from "app/providers/store";
import i18nTest from 'shared/config/i18n/i18nTest'
import { I18nextProvider } from "react-i18next"
import { MemoryRouter } from "react-router-dom";
import { DeepPartial, EnhancedStore } from "@reduxjs/toolkit";
import { AppThemeProvider } from "app/providers/theme";

export interface componentRenderOptions {
    route?: string;
    initialState?: DeepPartial<IStateSchema>;
    store?: EnhancedStore
}

const testWrapper = (component: React.ReactNode, options: componentRenderOptions = {}) => {
    const {
        route = '/',
        initialState,
        store
    } = options

    return render(
        <TestStoreProvider initialState={initialState} store={store}>
            <AppThemeProvider>
                <MemoryRouter initialEntries={[route]}>
                    <I18nextProvider i18n={i18nTest}>
                        {component}
                    </I18nextProvider>
                </MemoryRouter>
            </AppThemeProvider>
        </TestStoreProvider>
    )
}

export default testWrapper