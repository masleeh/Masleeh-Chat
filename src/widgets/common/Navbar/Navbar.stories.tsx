import { type Meta, type StoryObj } from '@storybook/react';
import Navbar from './Navbar'
import Decorator from 'shared/config/storybook/decorator';
import { Container } from '@mui/material';

const meta = {
    title: 'widgets/common/Navbar',
    component: Navbar,
    parameters: {
        layout: 'centered',
    },
    tags: ['autodocs'],
    argTypes: {

    },
} satisfies Meta<typeof Navbar>;

export default meta;
type Story = StoryObj<typeof meta>;

export const LoggedOut: Story = {
    args: {
        
    },
    decorators: [
        (StoryComponent) => (
            <Container maxWidth="md">
                <Decorator>
                    <StoryComponent />
                </Decorator>
            </Container>
        )
    ]
};

const state = {
    user: {
        userData: {
            user_id: 'asdfghjklzxv',
            username: 'Matvey Simonenkov',
            profile_pic: ''
        }
    }
}

export const LoggedIn: Story = {
    decorators: [
        (StoryComponent) => (
            <Decorator initialState={state}>
                <Container maxWidth="md">
                    <StoryComponent />
                </Container>
            </Decorator>
        )
    ]
}