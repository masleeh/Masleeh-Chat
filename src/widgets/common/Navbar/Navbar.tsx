import { AppBar, Container, Toolbar, Typography } from "@mui/material"
import AccountMenu from "./elements/AccountMenu/AccountMenu"
import { useSelector } from "react-redux"
import { getUserId } from "entities/User/model/selectors/getUserId/getUserId"
import useNavMenuButtons from "./utils/useNavMenuButtons"

const Navbar = () => {
    const userId = useSelector(getUserId)
    const {
        goToConversations
    } = useNavMenuButtons()
    
    return (
        <Container maxWidth="lg">
            <AppBar position="static" sx={{
                borderBottomLeftRadius: 15,
                borderBottomRightRadius: 15
            }}>
                <Toolbar>
                    <Typography 
                        variant="h5" 
                        component="div" 
                        sx={{ flexGrow: 1, cursor: 'pointer' }}
                        onClick={goToConversations}
                    >
                        Masleeh Chat
                    </Typography>

                    { userId && <AccountMenu />}
                            
                </Toolbar>
            </AppBar>
        </Container>
    )
}

export default Navbar