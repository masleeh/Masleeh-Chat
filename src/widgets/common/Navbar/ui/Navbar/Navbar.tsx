import { AppBar, Container, Toolbar, Typography } from "@mui/material"
import AccountMenu from "../AccountMenu/AccountMenu"
import { useSelector } from "react-redux"
import { getUserId } from "entities/User/model/selectors/getUserId/getUserId"
import useNavMenuButtons from "../../lib/useNavMenuButton/useNavMenuButtons"
import ToggleThemeMode from "../ToggleThemeMode/ToggleThemeMode"

const Navbar = () => {
    const userId = useSelector(getUserId)
    const {
        goToConversations
    } = useNavMenuButtons()
    
    return (
        <Container maxWidth="lg">
            <AppBar position="static" sx={{
                borderBottomLeftRadius: 15,
                borderBottomRightRadius: 15
            }}>
                <Toolbar>
                    <Typography 
                        variant="h5" 
                        component="div" 
                        sx={{ flexGrow: 1, cursor: 'pointer' }}
                        onClick={goToConversations}
                    >
                        Masleeh Chat
                    </Typography>

                    <ToggleThemeMode />
                    { userId && <AccountMenu sx={{ml: 2}}/>}
                            
                </Toolbar>
            </AppBar>
        </Container>
    )
}

export default Navbar