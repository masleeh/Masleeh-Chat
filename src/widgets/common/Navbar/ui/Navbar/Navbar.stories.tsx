import { type Meta, type StoryObj } from '@storybook/react';
import Navbar from './Navbar'
import Decorator from 'shared/config/storybook/decorator';
import { Container } from '@mui/material';
import { IStateSchema } from 'app/providers/store';

const meta = {
    title: 'widgets/common/Navbar',
    component: Navbar,
    parameters: {
        layout: 'centered',
    },
    tags: ['autodocs'],
    argTypes: {

    },
} satisfies Meta<typeof Navbar>;

export default meta;
type Story = StoryObj<typeof meta>;

export const LoggedOut: Story = {
    args: {
        
    },
    decorators: [
        (StoryComponent) => (
            <Container maxWidth="md">
                <Decorator>
                    <StoryComponent />
                </Decorator>
            </Container>
        )
    ]
};

const state: IStateSchema = {
    user: {
        userData: {
            user_id: 'asdfghjklzxv',
            username: 'Matvey Simonenkov',
            profile_pic: ''
        }
    },
    theme: {}
}

export const LoggedIn: Story = {
    decorators: [
        (StoryComponent) => (
            <Decorator initialState={state}>
                <Container maxWidth="md">
                    <StoryComponent />
                </Container>
            </Decorator>
        )
    ]
}