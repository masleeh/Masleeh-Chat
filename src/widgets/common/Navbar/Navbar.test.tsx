import testWrapper from "shared/lib/testWrapper/testWrapper"
import { screen } from '@testing-library/react'
import Navbar from "./Navbar"
import userEvent from "@testing-library/user-event"
import $api from "shared/lib/axiosApi/axiosApi"

const filledState = {
    user: {
        userData: {
            username: 'masleeh',
            profile_pic: 'localhost:5000/pics/7p3YxyvZruvN',
            user_id: 'asdfgghhjkk'
        }
    }
}

const emptyState = {
    user: {}
}

const axiosMock = vi.spyOn($api, 'post')

describe('Navbar component', () => {
    it('Check component', () => {
        const component = testWrapper(<Navbar />)
        expect(component).toMatchSnapshot()
    })
    
    it('Check avatar with filled state', () => {
        testWrapper(<Navbar />, {initialState: filledState})
        const AccountMenu = screen.getByLabelText('icon-button with avatar')
        expect(AccountMenu).toBeInTheDocument()
        expect(AccountMenu).toBeVisible()
    })
    
    it('Check avatar with empty state', () => {
        testWrapper(<Navbar />, {initialState: emptyState})
        const AccountMenu = screen.queryByLabelText('icon-button with avatar')
        expect(AccountMenu).not.toBeInTheDocument()
    })

    it('Check does logout removes userData', async () => {
        axiosMock.mockResolvedValue(() => vi.fn())

        testWrapper(<Navbar />, { initialState: filledState })
        const AvatarBtn = screen.getByLabelText('icon-button with avatar')
        await userEvent.click(AvatarBtn)
        const AccountMenu = screen.getByLabelText('navbar-menu')
        expect(AccountMenu).toBeInTheDocument()
        expect(AccountMenu).toBeVisible()
        
        const LogOutBtn = screen.getByLabelText("logout button")
        await userEvent.click(LogOutBtn)
        expect(AccountMenu).not.toBeVisible()
        expect(AccountMenu).not.toBeInTheDocument()

        expect(AvatarBtn).not.toBeInTheDocument()
    })
})