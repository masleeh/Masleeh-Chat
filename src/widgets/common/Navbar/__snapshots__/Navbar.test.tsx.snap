// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar component > Check component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-4dkfap-MuiPaper-root-MuiAppBar-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 css-1cs5hlo-MuiTypography-root"
            >
              Masleeh Chat
            </div>
          </div>
        </header>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-4dkfap-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 css-1cs5hlo-MuiTypography-root"
          >
            Masleeh Chat
          </div>
        </div>
      </header>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
